core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -55,22 +58,35 @@ private[kafka] abstract class SubSourceLogic[K, V, Msg](
   }
 
   val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
-    buffer = buffer.enqueue(tps.toList)
+    pendingPartitions ++= tps.filter(!partitionsInStartup.contains(_))
     pump()
   }
-  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { r =>
-    r.map(subSources.get).foreach(_.foreach(_.shutdown()))
-    subSources --= r
+
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
+    pendingPartitions --= tps
+    partitionsInStartup --= tps
+    tps.flatMap(subSources.get).foreach(_.shutdown())
+    subSources --= tps
   }
 
   val subsourceCancelledCB = getAsyncCallback[TopicPartition] { tp =>
     subSources -= tp
-    buffer :+= tp
+    partitionsInStartup -= tp
+    pendingPartitions += tp
```

Comment:
I was actually just copying what the code did before.  pendingPartitions is effectively the same thing as "buffer" used to be, but it is a Set instead of a Queue to prevent processing the same TopicPartition twice under conditions when partitions are revoked and reassigned quickly.

The old code put the TopicPartition back in buffer in this case, so I figured I should do the same with pendingPartitions to preserve the same behavior.